javaslang/src-gen/test/java/javaslang/CheckedFunction2Test.java

```diff
@@ -165,15 +165,6 @@ public void shouldLiftTryPartialFunction() {
         assertThat(unknown.getCause().getMessage()).isEqualToIgnoringCase("Unknown MessageDigest not available");
     }
 
-    private static final CheckedFunction2<Integer, Integer, Integer> recurrent1 = (i1, i2) -> i1 <= 0 ? i1 : CheckedFunction2Test.recurrent2.apply(i1 - 1, i2) + 1;
```

Comment:
@ruslansennov, @danieldietrich, @zsolt-donca, do you think something like
```java
final Map<Tuple3<T1, T2, T3>, R> cache = new HashMap<>();
return (Function3<T1, T2, T3, R> & Memoized) (t1, t2, t3) -> {
    final Tuple3<T1, T2, T3> key = Tuple.of(t1, t2, t3);
    if (cache.containsKey(key)) {
        return cache.get(key);
    } else {
        synchronized (cache) {
            if (cache.containsKey(key)) {
                return cache.get(key);
            } else {
                final R value = apply(t1, t2, t3);
                cache.put(key, value);
                return value;
            }
        }
    }
};
```
would work?

Seems to work correctly for all (recursive) tests, even on `Java 9`, but I don't think we have a complex enough test suite to be sure.
Also,  it's a pity that we cannot use `ConcurrentHashMap` because of the `non-null` value restriction.